### DISCUSSION TITLE
Custom Field Mutations

### DISCUSSION THREAD

**Discussion by zngly-vlad** [wpengine/atlas-content-modeler#514] [Ideas]
As a developer, I would like to be able to have mutations for the custom fields that I set up. This would make it easier for me to create my model and avoid writing boilerplate code in order to get each field appear in the mutation.

Currently only standard WordPress fields appear as part of mutations. 

The feature would have all (maybe enable mutation per field) fields to be mutated and appear in the mutation.

--- extra info

Having played with ACM and looking at the promising nature of its development, I believe that this missing feature is fundamental for developers making WordPress/Wpgraphql/Next websites and would also steer developers this way. 

My own personal needs stem from using [Wpgraphql](https://github.com/wp-graphql) with [Wpgraphql-acf](https://github.com/wp-graphql/wp-graphql-acf) . I needed to make quite a few mutations happen for the acf fields. After the mutations were implemented, I then created a plugin for [wpgraphql-acf-mutations](https://github.com/zngly/wp-graphql-acf-mutations) which automatically created mutations for all (most of) the acf fields.

--- tips

While making the plugin, I realised that having inputTypes were important and should have aggregated all the custom fields into an InputType. This would have helped with a number of issues described there and leads me to think thats the cleanest way to implement the feature. https://github.com/zngly/wp-graphql-acf-mutations/issues/1
The benefits are summarised below.

- reduce bloat input fields
- point all the acf inputs to a single entry point: acf
- avoid duplicate acf field key and original input keys

--- final note

This feature would allow me to start a new project using ACM and future developers to know they have the ability to fully control the data using graphql.

**Comment by nickcernis** [comment]
Thank you for the suggestion, @ditto-vlad! Some good news — we haven't released it yet, but we recently merged support for ACM fields as input fields for mutations in this PR:

https://github.com/wpengine/atlas-content-modeler/pull/499

Each field is registered as a separate input, consistent with existing WPGraphQL fields (rather than namespacing under an `acm` input field). We already have duplicate field slug detection so it's hard to create a field that uses the name of an existing field.

We plan to release this in our next update, likely this week or soon after.

You'll then be able to follow the docs here to make mutations on ACM models with ACM fields:

https://github.com/wpengine/atlas-content-modeler/blob/main/docs/mutations/index.md

**Reply by zngly-vlad** [reply]
That is great news to hear. Cant wait to try it out when it gets released!

**Reply by mindctrl** [reply]
@ditto-vlad ACM 0.17.0 is live with mutation support. Let us know what you think!

**Reply by zngly-vlad** [reply]
@mindctrl The mutations support is awesome! I assume more field types support are coming in the future (Media Field). Now you can truly have control over the data, query and mutate the models you define. I love it!

**Reply by nickcernis** [reply]
Thanks for trying it out and for the feedback, @ditto-vlad!

We hope to extend support to media and relationship fields later.

**Reply by zngly-vlad** [reply]
@nickcernis  Regarding the media field, given that it simply chooses from the media library in the wordpress ui, the graphql equivalent could just be a mediaItem id, you can leave actual file uploading for the devs.
